js/ccf-app/src/global.ts

```diff
@@ -312,6 +348,40 @@ export interface CCF {
    * Only defined for endpoints with "mode" set to "historical".
    */
   historicalState?: HistoricalState;
+
+  historical: {
+    /**
+     * Retrieve a range of historical states containing the state written at the given
+     * indices.
+     *
+     * If this is not currently available, this function returns `null`
+     * and begins fetching the ledger entry asynchronously. This will generally
+     * be true for the first call for a given seqno, and it may take some time
+     * to completely fetch and validate. The call should be repeated later with
+     * the same arguments to retrieve the requested entries. This state is kept
+     * until it is deleted for one of the following reasons:
+     *  - A call to {@link CCF.historical.dropCachedStateRange}
+     *  - `seconds_until_expiry` seconds elapse without calling this function
+     *  - This handle is used to request a different seqno or range
+     *
+     * The range is inclusive of both start_seqno and end_seqno. If a non-empty
+     * array is returned, it will always contain the full requested range; the
+     * array will be of length (end_seqno - start_seqno + 1).
+     */
+    getStateRange(
```

Comment:
I wasn't able to figure out from the c++ code (`get_state_range_internal` in `historical_queries.h`) what happens when unknown seqnos are passed. It seems like the request is simply passed on to the host, so the result will always be empty since it will never finish? @eddyashton

Reply:
The historical queries system doesn't check that the passed seqnos are known/committed. It could, but it's assumed this is better handled in the app. A request for a seqno which doesn't exist yet will be passed to the host, which will `consensus::ledger_no_entry` response message from the host, and the historical queries system deleting the state for this request (`handle_no_entry`).

Reply:
Right, but it also hides other error cases, like missing ledger files. In this case, being able to log the error and return a 500 or similar to the client would probably be better.

Reply:
Depending on the circumstances, a 404 is perhaps more accurate.